File: src/sw/sa_pipeline/stage5_sa.py
from src.util.util import Util


class Stage5SA:
    """
    Fifth Pipe from SA_Verilog. This pipe is responsible to find the manhandle
    distances for each combination between cellA and cellB with their 
    respective neighbors cells after swap and execute the first 2-2 additions
    for the distances found in the left pipe.
    """

    def __init__(self, n_lines: int, n_columns: int):
        """

        """
        self.n_lines: int = n_lines
        self.n_columns: int = n_columns
        self.new_output: dict = {
            'th_idx': 0,
            'th_valid': False,
            'dvac': [0, 0],
            'dvbc': [0, 0],
            'dvas': [0, 0, 0, 0],
            'dvbs': [0, 0, 0, 0]
        }
        self.old_output: dict = self.new_output.copy()

    def compute(self, st4_input: dict):
        """

        @param st4_input:
        """
        # moving forward the ready outputs
        self.old_output = self.new_output.copy()

        st4_th_idx: int = st4_input['th_idx']
        st4_th_valid: bool = st4_input['th_valid']

        st4_cbs: int = st4_input['cell_a']
        st4_cas: int = st4_input['cell_b']
        st4_cva: list = st4_input['cva']
        st4_cvb: list = st4_input['cvb']
        st4_dvac: list = st4_input['dvac']
        st4_dvbc: list = st4_input['dvbc']

        # fixme only for debugging
        if st4_th_idx == 0:
            z = 1

        dvac: list = [st4_dvac[0] + st4_dvac[1], st4_dvac[2] + st4_dvac[3]]
        dvbc: list = [st4_dvbc[0] + st4_dvbc[1], st4_dvbc[2] + st4_dvbc[3]]

        dvas = [0, 0, 0, 0]
        dvbs = [0, 0, 0, 0]

        for i in range(len(st4_cva)):
            if st4_cva[i] is not None:
                if st4_cas == st4_cva[i]:
                    dvas[i] = Util.dist_manhattan(
                        Util.get_line_column_from_cell(st4_cas, self.n_lines, self.n_columns),
                        Util.get_line_column_from_cell(st4_cbs, self.n_lines, self.n_columns)
                    )
                else:
                    dvas[i] = Util.dist_manhattan(
                        Util.get_line_column_from_cell(st4_cas, self.n_lines, self.n_columns),
                        Util.get_line_column_from_cell(st4_cva[i], self.n_lines, self.n_columns)
                    )

        for i in range(len(st4_cvb)):
            if st4_cvb[i] is not None:
                if st4_cbs == st4_cvb[i]:
                    dvbs[i] = Util.dist_manhattan(
                        Util.get_line_column_from_cell(st4_cas, self.n_lines, self.n_columns),
                        Util.get_line_column_from_cell(st4_cbs, self.n_lines, self.n_columns)
                    )
                else:
                    dvbs[i] = Util.dist_manhattan(
                        Util.get_line_column_from_cell(st4_cbs, self.n_lines, self.n_columns),
                        Util.get_line_column_from_cell(st4_cvb[i], self.n_lines, self.n_columns)
                    )

        self.new_output: dict = {
            'th_idx': st4_th_idx,
            'th_valid': st4_th_valid,
            'dvac': dvac,
            'dvbc': dvbc,
            'dvas': dvas,
            'dvbs': dvbs
        }

File: src/cpp/sw/sa_pipeline/stage5_sa.cpp
#include "sa_pipeline_sw.h"


class Stage5SA {
private:

public:
    ST5_OUT new_output = {
            0, false,
            {0, 0},
            {0, 0},
            {0, 0, 0, 0},
            {0, 0, 0, 0}
    };
    ST5_OUT old_output = {
            0, false,
            {0, 0},
            {0, 0},
            {0, 0, 0, 0},
            {0, 0, 0, 0}
    };

    void compute(ST4_OUT st4_input) {
        this->old_output = this->new_output;
        this->old_output = this->new_output;
        memcpy(&this->old_output.dvac, &this->new_output.dvac, sizeof(this->new_output.dvac));
        memcpy(&this->old_output.dvbc, &this->new_output.dvbc, sizeof(this->new_output.dvbc));
        memcpy(&this->old_output.dvas, &this->new_output.dvas, sizeof(this->new_output.dvas));
        memcpy(&this->old_output.dvbs, &this->new_output.dvbs, sizeof(this->new_output.dvbs));

        int st4_th_idx = st4_input.th_idx;
        bool st4_th_valid = st4_input.th_valid;
        int st4_cbs = st4_input.cell_a;
        int st4_cas = st4_input.cell_b;


        if (st4_th_idx == 0 && st4_th_valid) {
            int a = 1;
        }

        int dvac[2] = {st4_input.dvac[0] + st4_input.dvac[1], st4_input.dvac[2] + st4_input.dvac[3]};
        int dvbc[2] = {st4_input.dvbc[0] + st4_input.dvbc[1], st4_input.dvbc[2] + st4_input.dvbc[3]};

        int dvas[N_NEIGH] = {0, 0, 0, 0};
        int dvbs[N_NEIGH] = {0, 0, 0, 0};

        for (int n = 0; n < N_NEIGH; ++n) {
            int i1, i2, j1, j2;

            if (st4_input.cva[n] != -1) {
                get_line_column_from_cell(st4_cas, N_LINES, N_COLUMNS, i1, j1);
                if (st4_cas == st4_input.cva[n]) {
                    get_line_column_from_cell(st4_cbs, N_LINES, N_COLUMNS, i2, j2);
                } else {
                    get_line_column_from_cell(st4_input.cva[n], N_LINES, N_COLUMNS, i2, j2);
                }
#if defined(ONE_HOP)
                dvas[n] = dist_one_hop(i1, j1, i2, j2);
#elif defined(MESH)
                dvas[n] = dist_manhattan(i1, j1, i2, j2);
#endif
            }

            if (st4_input.cvb[n] != -1) {
                get_line_column_from_cell(st4_cbs, N_LINES, N_COLUMNS, i1, j1);
                if (st4_cbs == st4_input.cvb[n]) {
                    get_line_column_from_cell(st4_cas, N_LINES, N_COLUMNS, i2, j2);
                } else {
                    get_line_column_from_cell(st4_input.cvb[n], N_LINES, N_COLUMNS, i2, j2);
                }
#if defined(ONE_HOP)
                dvbs[n] = dist_one_hop(i1, j1, i2, j2);
#elif defined(MESH)
                dvbs[n] = dist_manhattan(i1, j1, i2, j2);
#endif
            }
        }

        this->new_output.th_idx = st4_th_idx;
        this->new_output.th_valid = st4_th_valid;
        memcpy(&this->new_output.dvac, &dvac, sizeof(dvac));
        memcpy(&this->new_output.dvbc, &dvbc, sizeof(dvbc));
        memcpy(&this->new_output.dvas, &dvas, sizeof(dvas));
        memcpy(&this->new_output.dvbs, &dvbs, sizeof(dvbs));
    }
};